### ISSUE TITLE
MLP layer 분석

### ISSUE THREAD

**Issue by februa22** [nlp-research/biaffineparser#5]
목적: 알고리즘 분석을 하고 적절한 TensorFlow API를 찾는다.

MLP 레이어는 bidirectional LSTMs 레이어의 다음 레이어입니다.

**Comment by februa22** [comment]
Shape of output

- lstm: [100, 63, 900]
- arc-dep: [100, 63, 500]
- arc-head: [100, 63, 500]
- label-dep: [100, 63, 100]
- label-head: [100, 63, 100]

**Comment by februa22** [comment]
Shape of output

- lstm: [batch_size, seq_len, 2*lstm_hidden_size]
- arc-dep: [batch_size, seq_len, arc_mlp_units]
- arc-head: [batch_size, seq_len, arc_mlp_units]
- label-dep: [batch_size, seq_len, label_mlp_units]
- label-head: [batch_size, seq_len, label_mlp_units]

**Comment by februa22** [comment]
MLP layer 전에 stacked bi-lstm은 다음 함수로 구현 가능:

```python
def add_stacked_lstm_layers(hparams, word_embedding, lengths):
    cell = tf.contrib.rnn.LSTMCell
    cells_fw = [cell(hparams.lstm_hidden_size) for _ in range(hparams.num_layers)]
    cells_bw = [cell(hparams.lstm_hidden_size) for _ in range(hparams.num_layers)]
    if hparams.dropout > 0.0:
        cells_fw = [tf.contrib.rnn.DropoutWrapper(cell) for cell in cells_fw]
        cells_bw = [tf.contrib.rnn.DropoutWrapper(cell) for cell in cells_bw]
    outputs, _, _ = tf.contrib.rnn.stack_bidirectional_dynamic_rnn(
        cells_fw=cells_fw,
        cells_bw=cells_bw,
        inputs=word_embedding,
        sequence_length=lengths,
        dtype=tf.float32,
        scope="bi-lstm")
    return outputs
```

**Comment by februa22** [comment]
MLP layer 구현 완료.

함수명: `mlp_for_arc_and_label`

```python
def create_weight_and_bias(n_input, n_output):
    weights = {
        'w1': tf.get_variable('w1', shape=[n_input, n_output], dtype=tf.float32),
        'w2': tf.get_variable('w2', shape=[n_output, n_output], dtype=tf.float32),
    }
    biases = {
        'b1': tf.get_variable('b1', shape=[n_output], dtype=tf.float32, initializer=tf.zeros_initializer()),
        'b2': tf.get_variable('b2', shape=[n_output], dtype=tf.float32, initializer=tf.zeros_initializer()),
    }
    return weights, biases


def multilayer_perceptron(x, weights, biases):
    layer_1 = tf.add(tf.matmul(x, weights['w1']), biases['b1'])
    layer_1 = tf.nn.relu(layer_1)
    layer_2 = tf.add(tf.matmul(layer_1, weights['w2']), biases['b2'])
    layer_2 = tf.nn.relu(layer_2)
    return layer_2


def mlp_with_scope(x, n_input, n_output, scope):
    with tf.variable_scope(scope):
        weights, biases = create_weight_and_bias(n_input, n_output)
        h = multilayer_perceptron(x, weights, biases)
    return h


def mlp_for_arc(hparams, x):
    h_arc_head = mlp_with_scope(x, 2*hparams.lstm_hidden_sizem, hparams.arc_head_units, 'arc_head')
    h_arc_dep = mlp_with_scope(x, 2*hparams.lstm_hidden_sizem, hparams.arc_dep_units, 'arc_dep')
    return h_arc_head, h_arc_dep


def mlp_for_label(hparams, x):
    h_label_head = mlp_with_scope(x, 2*hparams.lstm_hidden_sizem, hparams.label_head_units, 'label_head')
    h_label_dep = mlp_with_scope(x, 2*hparams.lstm_hidden_sizem, hparams.label_dep_units, 'label_dep')
    return h_label_head, h_label_dep


def mlp_for_arc_and_label(hparams, x):
    x = tf.nn.dropout(x, hparams.mlp_dropout)
    h_arc_head, h_arc_dep = mlp_for_arc(hparams, x)
    h_label_head, h_label_dep = mlp_for_label(hparams, x)
    return h_arc_head, h_arc_dep, h_label_head, h_label_dep
```

**Comment by februa22** [comment]
`parser/model.py`에 구현 완료.
